xcm/pallet-xcm/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+use xcm::v0::{BodyId, MultiLocation::{self, X1}, Junction::Plurality};
+use sp_runtime::{RuntimeDebug, traits::BadOrigin};
+use frame_support::traits::{EnsureOrigin, OriginTrait, Filter, Get};
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use xcm::v0::{Xcm, MultiLocation, Error as XcmError, SendXcm, ExecuteXcm};
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Required origin for sending XCM messages. If successful, the it resolves to `MultiLocation`
+		/// which exists as an interior location within this chain's XCM context.
+		type SendXcmOrigin: EnsureOrigin<Self::Origin, Success=MultiLocation>;
+
+		/// The type used to actually dispatch an XCM to its destination.
+		type XcmRouter: SendXcm;
+
+		/// Required origin for executing XCM messages. If successful, the it resolves to `MultiLocation`
+		/// which exists as an interior location within this chain's XCM context.
+		type ExecuteXcmOrigin: EnsureOrigin<Self::Origin, Success=MultiLocation>;
+
+		/// Something to execute an XCM message.
+		type XcmExecutor: ExecuteXcm<Self::Call>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Attempted(xcm::v0::Outcome),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		Unreachable,
+		SendFailure,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(1_000)]
+		fn send(origin: OriginFor<T>, dest: MultiLocation, message: Xcm<()>) -> DispatchResult {
+			let origin_location = T::SendXcmOrigin::ensure_origin(origin)?;
+			Self::send_xcm(origin_location, dest, message)
+				.map_err(|e| match e {
+					XcmError::CannotReachDestination(..) => Error::<T>::Unreachable,
+					_ => Error::<T>::SendFailure,
+				})?;
+			Ok(())
+		}
+
+		/// Execute an XCM message from a local, signed, origin.
+		///
+		/// Returns `DispatchError` if execution was not possible. Otherwise returns the amount of
+		/// weight that was used in its attempted execution.
+		///
+		/// An event is deposited indicating whether `msg` could be executed completely or only
+		/// partially.
+		///
+		/// No more than `max_weight` will be used in its attempted execution. If this is less than the
+		/// maximum amount of weight that the message could take to be executed, then no execution
+		/// attempt will be made.
+		///
+		/// NOTE: A successful return to this does *not* imply that the `msg` was executed successfully
+		/// to completion; only that *some* of it was executed.
+		#[pallet::weight(1_000 + max_weight)]
```

Comment:
```suggestion
		#[pallet::weight(1_000.saturating_add(max_weight))]
```

My understanding is this calculation can still be triggered by an external user who does not match this origin in the transaction queue.